### PR TITLE
Create monitor to log events remotely

### PR DESCRIPTION
**Pull request by drewprice**

### CODE REVIEW

File: lib/command-log.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+'use babel'
```
Comment (joshrowley) [inline lib/command-log.js]: atom has babel support?

Comment (joshrowley) [inline lib/command-log.js]: also if we're switching to using es6 imports etc...can we apply the same style to the rest of the codebase as well. just want to have it match one way or antoher

File: lib/http.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
+'use babel'
+
+import token from './token'
+
+function httpRequest(method, path, body, learnAuth=false) {
```
Comment (joshrowley) [inline lib/http.js]: at some point i'd like to reconcile our JS http stuff into one library we use across the board, either a 3rd party one we use as a standard, or our own implementation / wrapper

File: lib/log.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
+'use babel'
+
+import commandLog from './command-log'
+import {learnCo} from './config'
+import {post, patch} from './http'
+import token from './token'
+import version from './version'
+
+export default {
+  defaultData() {
```
Comment (joshrowley) [inline lib/log.js]: prob don't need this as part of the public interface (could be private fn inside hidden in this module

File: lib/monitor.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  return LEARN_IDE_HOST_IP
+}
+
+function clearHostIP() {
+  LEARN_IDE_HOST_IP = null
+}
+
+export default function monitor(terminal, subscriptions) {
+  subscriptions.add(atom.commands.onDidDispatch(({type}) => {
+    if (type === 'application:quit') { log.add({event: 'quit'}) }
```
Comment (joshrowley) [inline lib/monitor.js]: is it ever possible that the application quits before the post is made / can we guarantee this executes before the program exits?

File: lib/terminal.coffee [hunk 5]
```diff
@@ -46,9 +50,13 @@ module.exports = class Terminal extends EventEmitter
     "#{protocol}://#{@host}:#{@port}/#{@path}?token=#{@token}&version=#{version}"
 
   reset: ->
-    console.log('term:reset')
     @socket.reset()
 
+  sendWithPromise: (msg, test) ->
```
Comment (joshrowley) [inline lib/terminal.coffee]: took me a sec to grok what this function did, wondering if there's a more detailed name we can give like `sendAndReceive` or `sendAndHandleReponse` (not tied to either of those name, but something a little more specific then `sendWithPromise`

**Comment by joshrowley** [comment]
looks good 👍

**Comment by joshrowley** [comment]
i would wait until monday to release though, how much QA has been done on it?

File: lib/command-log.js (hunk 1 above)
Comment (drewprice) [inline lib/command-log.js]: Yep yep, v2.5 has more stuff in es6, and I've been slowly moving things over with [decaffeinate](http://decaffeinate-project.org/). So it'll be a little mixed while we transition but soon enough everything will be es6.

File: lib/monitor.js (hunk 4 above)
Comment (drewprice) [inline lib/monitor.js]: I'm not sure that we can guarantee it; however, it's been hitting this really reliably. We can move it up a bit using `atom.commands.onWillDispatch`, though I dunno how much that actually buys us.

**Comment by drewprice** [comment]
@joshrowley just myself testing it, so it definitely needs some QA time before it's ready.

File: lib/http.js (hunk 2 above)
Comment (drewprice) [inline lib/http.js]: Yeah, it'd definitely be nice & would save a lot of time to have a standard throughout

File: lib/monitor.js (hunk 4 above)
Comment (drewprice) [inline lib/monitor.js]: worked reliably on mac... not so much on windows. so yeah, I need to figure out some way to try and guarantee this.

**Pull request merged**